app/products/calls/state/actions.ts

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {getCallsConfig, exportedForInternalUse as callsConfigInternal} from '@calls/state/calls_config';
+import {getCallsState, exportedForInternalUse as callsStateInternal} from '@calls/state/calls_state';
+import {getChannelsWithCalls, exportedForInternalUse as channelsWithCallsInternal} from '@calls/state/channels_with_calls';
+import {getCurrentCall, exportedForInternalUse as currentCallInternal} from '@calls/state/current_call';
+import {Call, ChannelsWithCalls, ServerConfig} from '@calls/types/calls';
+
+const {setCallsConfig} = callsConfigInternal;
+const {setCallsState} = callsStateInternal;
+const {setChannelsWithCalls} = channelsWithCallsInternal;
+const {setCurrentCall} = currentCallInternal;
+
+export const setCalls = (serverUrl: string, myUserId: string, calls: Dictionary<Call>, enabled: Dictionary<boolean>) => {
+    const channelsWithCalls = Object.keys(calls).reduce(
+        (accum, next) => {
+            accum[next] = true;
+            return accum;
+        }, {} as ChannelsWithCalls);
+    setChannelsWithCalls(serverUrl, channelsWithCalls);
+
+    setCallsState(serverUrl, {serverUrl, myUserId, calls, enabled});
+};
+
+export const userJoinedCall = (serverUrl: string, channelId: string, userId: string) => {
+    const callsState = getCallsState(serverUrl);
+    if (!callsState.calls[channelId]) {
+        return;
+    }
+
+    const nextCall = {
+        ...callsState.calls[channelId],
+        participants: {...callsState.calls[channelId].participants},
+    };
+    nextCall.participants[userId] = {
+        id: userId,
+        muted: true,
+        raisedHand: 0,
+    };
+    const nextCalls = {...callsState.calls, [channelId]: nextCall};
+
+    setCallsState(serverUrl, {...callsState, calls: nextCalls});
+
+    // Did the user join the current call? If so, update that too.
+    const currentCall = getCurrentCall();
+    if (!currentCall || currentCall.channelId !== channelId) {
+        return;
+    }
+
+    const nextCall2 = {
+        ...currentCall,
+        participants: {...currentCall.participants, [userId]: nextCall.participants[userId]},
+    };
+    setCurrentCall(nextCall2);
```

Comment:
Connecting to a previous comment, this will trigger the observable many times during a call. If you don't properly use the `distinctUntilChanged` in the pipes that come out of this, you may be re-rendering and re-triggering pipes too often.

Reply:
Yep, thank you! 50d02fc2d55999db71916081b71de08be97bdb9a